cutgeneratingfunctionology/spam/basic_semialgebraic.py

```diff
@@ -2230,3 +2276,714 @@ def tighten_upstairs_by_mccormick(self, max_iter=1):
+        r"""
+        Add the constraint ``lhs`` * x + cst ``op`` 0,
+        where ``lhs`` is a vector of length ambient_dim, and
+        ``op`` is one of ``operator.lt``, ``operator.gt``, ``operator.eq``,
+        ``operator.le``, ``operator.ge``
+        
+        EXAMPLES::
+        
+            sage: from cutgeneratingfunctionology.spam.basic_semialgebraic import *
+            sage: P = BasicSemialgebraicSet_polyhedral_Normaliz_semiopen_polyhedron(2)
+            sage: P.add_linear_constraint([2,3],-6,operator.ge)
+            sage: sorted(P.le_poly())
+            [-2*x0 - 3*x1 + 6]
+        """
+        # We store as the underlying polyhedron the closure of the given constrains. This let's us keep a (closed) double description.
+        self._Normaliz_constraint(lhs, cst, op)
+        if len(self._constraints["inhom_equations"]) == 0 :
+            self._polyhedron = Polyhedron_normaliz(self._ambient_space, Vrep = None, Hrep = [self._constraints["inhom_inequalities"]+self._constraints["inhom_excluded_faces"], None])
+        elif len(self._constraints["inhom_inequalities"]) == 0 :
+            self._polyhedron = Polyhedron_normaliz(self._ambient_space, Vrep = None, Hrep = [None, self._constraints["inhom_equations"]])
+        else:
+            self._polyhedron = Polyhedron_normaliz(self._ambient_space, Vrep = None, Hrep = [self._constraints["inhom_inequalities"]+self._constraints["inhom_excluded_faces"], self._constraints["inhom_equations"]])
+
+    def is_empty(self):
+        """
+        EXAMPLES::
+
+            sage: from cutgeneratingfunctionology.spam.basic_semialgebraic import *
+            sage: S = BasicSemialgebraicSet_polyhedral_Normaliz_semiopen_polyhedron(1)
+            sage: S.add_linear_constraint([1], -1, operator.ge)
+            sage: S.is_empty()
+            False
+            sage: S.add_linear_constraint([1], +1, operator.le)
+            sage: S.is_empty()
+            True
+        """
+        # by cones. 
+        return self._polyhedron._nmz_result(self._polyhedron._make_normaliz_cone(self._constraints), "IsEmptySemiOpen")
+        # return self._polyhedron.is_empty()
+        #return self._polyhedron.IsEmptySemiOpen()
+
+    def is_universe(self): 
+        """
+        EXAMPLES::
+
+            sage: from cutgeneratingfunctionology.spam.basic_semialgebraic import *
+            sage: S = BasicSemialgebraicSet_polyhedral_Normaliz_semiopen_polyhedron(1)
+            sage: S.add_linear_constraint([0], 0, operator.eq)
+            sage: S.is_universe()
+            True
+            sage: S.add_linear_constraint([1], +1, operator.le)
+            sage: S.is_universe()
+            False
+        """
+        # suppose that the semiopen constrtaints are the universe, then the closure of these constraints is also the universe, this command needs not modifications
+        return self._polyhedron.is_universe()
+
+
+class BasicSemialgebraicSet_polyhedral_pplite_NNC_Polyhedron(BasicSemialgebraicSet_polyhedral):
```

Comment:
I'd suggest moving this new class into a separate file `basic_semialgebraic_pplite`